File: ros2_ws/src/my_cpp_pkg/src/number_publisher.cpp
#include <chrono>
#include <string_view>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"

using msg_type = std_msgs::msg::Int64;
namespace {
std::string_view sv_name{"number_publisher"};
std::string_view topic_name{"number"};
}  // namespace

class NumberPublisherNode : public rclcpp::Node {
 public:
  NumberPublisherNode() : Node(sv_name.data()) {
    publisher_ = this->create_publisher<msg_type>(topic_name.data(), 10);
    timer_ =
        this->create_wall_timer(std::chrono::seconds(1),
                                std::bind(&NumberPublisherNode::publish, this));
    RCLCPP_INFO(this->get_logger(), "%s has started...", sv_name.data());
  }

 private:
  rclcpp::Publisher<msg_type>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  void publish(void) {
    static int64_t counter{};
    msg_type msg;
    msg.data = counter++;
    publisher_->publish(msg);
  }
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<NumberPublisherNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: ros2_ws/src/my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64

node_name = 'number_publisher'

class NumberPublisherNode(Node):
    def __init__(self):
        super().__init__(f'{node_name}')
        self.get_logger().info(f'{node_name} has started...')
        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.timer_ = self.create_timer(1, self.publishNumber)
        self.count = 1
    def publishNumber(self):
        msg = Int64()
        msg.data = self.count
        self.count += 1
        self.publisher_.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()